File: src/apps/scala/slate-examples/src/main/scala/slate/examples/Example_Database.scala
/**
<slate_header>
  author: Kishore Reddy
  url: https://github.com/kishorereddy/scala-slate
  copyright: 2015 Kishore Reddy
  license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  desc: a scala micro-framework
  usage: Please refer to license on github for more info.
</slate_header>
  */


package slate.examples

//<doc:import_required>
import slate.common.databases.{DbConfig, DbLookup, Db, DbConString}
import slate.common.results.ResultSupportIn
import slate.core.common.Conf
import slate.entities.core.EntityMapper
import slate.examples.common.User

//</doc:import_required>
//<doc:import_examples>
import slate.core.cmds.Cmd
import slate.common.{Result}
import scala.reflect.runtime.universe.{typeOf}
//</doc:import_examples>

class Example_Database extends Cmd("types") with ResultSupportIn {

  override protected def executeInternal(args: Any) : AnyRef =
  {
    //<doc:examples>
    // CASE 1: Default database connection using credentials from users folder with content:
    //
    // e.g. on windows: C:\Users\kv\slatekit\conf\db.txt
    //
    // driver:com.mysql.jdbc.Driver
    // url:jdbc:mysql://localhost/World
    // user:root
    // password:123abc
    DbLookup.setDefault( Some(DbConfig.loadFromUserFolder(".slate", "db_default.txt") ))

    // CASE 2: Default database connection explicitly ( not recommended - store config files in your user folder )
    DbLookup.setDefault("com.mysql.jdbc.Driver", "jdbc:mysql://localhost/default", "root", "abcdefghi")
    showResult( DbLookup.getDefault )

    // CASE 3: Register connection and link to a key "user_db" using credentials from user folder
    DbLookup.set("user_db", DbConfig.loadFromUserFolder(".slate", "db_default.txt") )
    showResult( DbLookup.getByKey("user_db") )

    // CASE 4: Register connection and link to a key "user_db" explicitly
    // ( not recommended - store config files in your user folder )
    DbLookup.set("user_db", "com.mysql.jdbc.Driver", "jdbc:mysql://localhost/user_db", "root", "abcdefghi")
    showResult( DbLookup.getByKey("user_db") )

    // CASE 5: Register connection as a shard and link to a group
    DbLookup.shard("group1", "shard1", DbConfig.loadFromUserFolder(".slate", "db_default.txt") )
    DbLookup.shard("group1", "shard2", "com.mysql.jdbc.Driver", "jdbc:mysql://localhost/World_shard2", "root", "abcdefghi")
    showResult( DbLookup.getShard("group1", "shard1") )
    showResult( DbLookup.getShard("group1", "shard2") )

    // CASE 6: Now create the database class and start using some basic methods
    val db = new Db(new Conf().dbCon("db").get)

    // CASE 7: Open the database
    db.open()

    // CASE 8: Get scalar int value
    val total1 = db.getScalarInt("select count(*) from City")

    // CASE 9: Execute a sql insert
    val id1 = db.executeInsertGetId("insert into `city`(`name`) values( 'ny' )" )

    // CASE 10: Execute a sql insert using
    val id2 = db.executeInsertGetId("insert into `city`(`name`) values( 'ny' )" )

    // CASE 10: Execute a sql update
    val total2 = db.executeUpdate("update `city` set `name` = 'ny' where id = 2")

    // CASE 11: Map a record to an model using the mapper component
    val mapper = new EntityMapper(null)
    mapper.loadSchema(new User(), typeOf[User])
    val model = db.mapOne("select * from `city` where id = 1", mapper)
    println( model )

    // CASE 12: Map multiple records
    val models = db.mapMany("select * from `city` where id < 5", mapper)
    println( models )

    //</doc:examples>
    ok()
  }


  def showResult(con: DbConString):Unit =
  {
    println(con)
  }
}

File: src/apps/kotlin/slatekit-examples/src/main/kotlin/slatekit/examples/Example_Database.kt
/**
<slate_header>
  author: Kishore Reddy
  url: https://github.com/kishorereddy/scala-slate
  copyright: 2015 Kishore Reddy
  license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  desc: a scala micro-framework
  usage: Please refer to license on github for more info.
</slate_header>
  */


package slatekit.examples

//<doc:import_required>
import slatekit.common.db.*
import slatekit.common.db.DbLookup.DbLookupCompanion.defaultDb
import slatekit.common.db.DbLookup.DbLookupCompanion.namedDbs
import slatekit.common.mapper.Mapper
import slatekit.entities.core.EntityMapper

//</doc:import_required>

//<doc:import_examples>
import slatekit.core.cmds.Cmd
import slatekit.common.Result
import slatekit.common.conf.Config
import slatekit.common.results.ResultFuncs.ok
import slatekit.examples.common.User

//</doc:import_examples>

class Example_Database : Cmd("db") {

  override fun executeInternal(args: Array<String>?) : Result<Any>
  {
    //<doc:examples>
    // These examples show 2 parts of the database component:
    // 1. Database connection registration
    // 2. Database reading/writing

    // Manual creation via class
    val dbs1 = DbLookup()
    showResult( dbs1.default() )

    // CASE 2: Default database connection using credentials from users folder with content:
    // e.g. on windows: C:\Users\kv\slatekit\conf\db.txt
    //
    // driver:com.mysql.jdbc.Driver
    // url:jdbc:mysql://localhost/World
    // user:root
    // password:123abc
    val dbs2 = defaultDb( DbUtils.loadFromUserFolder(".slatekit", "db_default.txt"))
    showResult( dbs2.default() )

    // CASE 3: Default database connection explicitly
    val dbs3 = defaultDb(
      DbConString (
        "com.mysql.jdbc.Driver",
        "jdbc:mysql://localhost/default",
        "root",
        "abcdefghi"
      )
    )
    showResult( dbs3.default() )

    // CASE 4: Register connection and link to a key "user_db" using credentials from user folder
    val dbs4 = namedDbs( listOf(Pair("users", DbUtils.loadFromUserFolder(".slate", "db_default.txt"))))
    showResult( dbs4.named("users") )

    // CASE 5: Register multiple connections by api ( "users", "files" )
    val dbs5 = namedDbs(listOf(
      Pair(
        "users", DbConString (
          "com.mysql.jdbc.Driver",
          "jdbc:mysql://localhost/users",
          "root",
          "abcdefghi"
        )
      ),
      Pair(
        "files", DbConString (
          "com.mysql.jdbc.Driver",
          "jdbc:mysql://localhost/files",
          "root",
          "abcdefghi"
        )
      )
    ))
    showResult( dbs5.named("users") )
    showResult( dbs5.named("fiels") )



    // CASE 7: Now create the database class and start using some basic methods
    val db = Db(Config().dbCon("db"))

    // CASE 8: Open the database
    db.open()

    // CASE 8: Get scalar int value
    val total1 = db.getScalarInt("select count(*) from City")

    // CASE 9: Execute a sql insert
    val id1 = db.insert("insert into `city`(`api`) values( 'ny' )" )

    // CASE 10: Execute a sql insert using
    val id2 = db.insert("insert into `city`(`api`) values( 'ny' )" )

    // CASE 10: Execute a sql update
    val total2 = db.update("update `city` set `api` = 'ny' where id = 2")

    // CASE 11: Map a record to an model using the mapper component
    val model = Mapper.loadSchema(User::class)
    val mapper = EntityMapper(model)
    val item1 = db.mapOne<User>("select * from `city` where id = 1", mapper)
    println( item1 )

    // CASE 12: Map multiple records
    val items = db.mapMany<User>("select * from `city` where id < 5", mapper)
    println( items )

    //</doc:examples>
    return ok()
  }


  fun showResult(con: DbCon?):Unit {
    println(con)
  }
}
